DFG should support tuples

kmiller68 opened the pull request:
#### f2f3c91fa89c2bef7d30bba0324cdba55aa3ae4f
<pre>
DFG should support tuples
<a href="https://bugs.webkit.org/show_bug.cgi?id=253413">https://bugs.webkit.org/show_bug.cgi?id=253413</a>

Reviewed by Yusuke Suzuki.

This change adds support for tuples in the DFG. It works similarly to how tuples
work in B3 where there is a side buffer which holds most of the metadata for a
tuple. In the DFG there are three pieces of information held here:
    1) The reference count
    2) The result flags
    3) The virtual register

1) tells us if the ExtractFromTuple Node for a given tuple result still exists
(e.g. it could have been constant folded or dead code eliminated). From that we
will decide to fill (3) the virtual register when allocating registers. If we
didn&apos;t have the reference count then we would have no way to know that the
virtual registers isn&apos;t going to be consumed thus we will leak it. Lastly we
have (2) the result flags for the node. This tells the ExtractFromTuple and
theoretically (as right now the only ExtractFromTuple users produce Int32s) any
 consumers of the ExtractFromTuple what value format the tuple&apos;s result will
 be in. For the DFG, we don&apos;t need this because we know there&apos;s at most one
 ExtractFromTuple for any given tuple index since we don&apos;t duplicate code.

When dumping the DFG graph ExtractFromTuple follows the same pattern as B3
and uses `&lt;&lt;X` to denote the offset we are extracting from. So it will look
something like:

  6  1  0:  D@101:&lt; 1:-&gt;    EnumeratorNextUpdateIndexAndMode(Check:Untyped:D@97, Check:Untyped:D@98, Check:Untyped:D@100, Check:Untyped:D@99, VarArgs, SelectUsingPredictions+NonArray+InBounds+AsIs+Read, enumeratorModes = 4, R:World, W:Heap, Exits, ClobbersExit, bc#115, ExitValid)
  7  1  0:  D@102:&lt; 1:-&gt;    ExtractFromTuple(Check:Untyped:D@101, Int32|UseAsOther, &lt;&lt;0, bc#115, ExitInvalid)
  8  1  0:  D@103:&lt;!0:-&gt;    MovHint(Check:Untyped:D@102, MustGen, loc10, W:SideState, ClobbersExit, bc#115, ExitInvalid)
  9  1  0:  D@104:&lt; 1:-&gt;    ExtractFromTuple(Check:Untyped:D@101, Int32|UseAsOther, &lt;&lt;1, bc#115, ExitInvalid)

This patch also adds support for calling operations in both the FTL/B3 via
CCall. CCall can take exactly the tuple of `{ pointerType(), pointerType() }`,
which, for every calling conevention we support, should be returned in both
the return value registers. As the only way to look into a tuple is via the
B3 prodecure, the first Air::Arg of any CCall/ColdCCall Inst is now the
CCallSpecial for the compiling Air::Code. This gives us access to Air::Code
inside CCallCustom::forEachArg and isValidForm.

* JSTests/stress/for-in-redefine-enumerable.js:
(shouldBe):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Type.h:
(JSC::B3::pointerType):
(JSC::B3::registerType):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:
(JSC::B3::Air::cCallResultCount):
(JSC::B3::Air::cCallArgumentRegisterWidth):
(JSC::B3::Air::cCallResult):
* Source/JavaScriptCore/b3/air/AirCCallingConvention.h:
* Source/JavaScriptCore/b3/air/AirCustom.cpp:
(JSC::B3::Air::CCallCustom::isValidForm):
* Source/JavaScriptCore/b3/air/AirCustom.h:
(JSC::B3::Air::CCallCustom::forEachArg):
* Source/JavaScriptCore/b3/air/AirLowerAfterRegAlloc.cpp:
(JSC::B3::Air::lowerAfterRegAlloc):
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_3.cpp:
(addCallTests):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(JSC_DEFINE_JIT_OPERATION):
(testCallPairResult):
(testCallPairResultRare):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreter.h:
(JSC::DFG::AbstractInterpreter::setTupleConstant):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGAtTailAbstractState.cpp:
(JSC::DFG::AtTailAbstractState::AtTailAbstractState):
* Source/JavaScriptCore/dfg/DFGAtTailAbstractState.h:
(JSC::DFG::AtTailAbstractState::forTupleNode):
(JSC::DFG::AtTailAbstractState::clearForTupleNode):
(JSC::DFG::AtTailAbstractState::setForTupleNode):
(JSC::DFG::AtTailAbstractState::setTypeForTupleNode):
(JSC::DFG::AtTailAbstractState::setNonCellTypeForTupleNode):
(JSC::DFG::AtTailAbstractState::makeBytecodeTopForTupleNode):
(JSC::DFG::AtTailAbstractState::makeHeapTopForTupleNode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::addToGraph):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGenerationInfo.h:
(JSC::DFG::GenerationInfo::initFromTupleResult):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::dump):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.cpp:
(JSC::DFG::InPlaceAbstractState::InPlaceAbstractState):
* Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.h:
(JSC::DFG::InPlaceAbstractState::forTupleNode):
(JSC::DFG::InPlaceAbstractState::clearForTupleNode):
(JSC::DFG::InPlaceAbstractState::setForTupleNode):
(JSC::DFG::InPlaceAbstractState::setTypeForTupleNode):
(JSC::DFG::InPlaceAbstractState::setNonCellTypeForTupleNode):
(JSC::DFG::InPlaceAbstractState::makeBytecodeTopForTupleNode):
(JSC::DFG::InPlaceAbstractState::makeHeapTopForTupleNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::isTuple const):
(JSC::DFG::Node::setTupleOffset):
(JSC::DFG::Node::tupleOffset const):
(JSC::DFG::Node::hasExtractOffset const):
(JSC::DFG::Node::extractOffset const):
(JSC::DFG::Node::tupleIndex const):
(JSC::DFG::Node::tupleSize const):
(JSC::DFG::Node::hasVirtualRegister):
(JSC::DFG::Node::virtualRegister):
(JSC::DFG::Node::setVirtualRegister):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
(JSC::DFG::makeUGPRPair):
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGScoreBoard.h:
(JSC::DFG::ScoreBoard::use):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::strictInt32TupleResultWithoutUsingChildren):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGValidate.cpp:
* Source/JavaScriptCore/dfg/DFGVirtualRegisterAllocationPhase.cpp:
(JSC::DFG::VirtualRegisterAllocationPhase::run):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::LowerDFGToB3):
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::extract):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::emitCCall):

Canonical link: <a href="https://commits.webkit.org/262068@main">https://commits.webkit.org/262068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab3884c0eee7246644afa330096caa5fe6815207

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/345 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/377 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/328 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/533 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/570 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/378 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/330 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/400 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/317 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/307 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/347 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/357 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/331 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/365 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/299 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/342 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/49 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/109 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/335 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/364 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/49 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -388,6 +388,14 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
     }
+
+    case ExtractFromTuple: {
+
```
Remove this empty line.

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -4484,10 +4492,12 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         if (node->enumeratorMetadata() == JSPropertyNameEnumerator::OwnStructureMode && m_graph.varArgChild(node, 0).useKind() == CellUse) {
-            // Do nothing.
-        } else if (node->enumeratorMetadata() != JSPropertyNameEnumerator::IndexedMode)
+            setTupleConstant(node, 1, jsNumber(static_cast<uint8_t>(JSPropertyNameEnumerator::OwnStructureMode)));
```
Remove brace for one line statement.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,102 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
     flushRegisters();
     JSValueRegsFlushedCallResult result(this);
     JSValueRegs resultRegs = result.regs();
+    // FIXME: We should teach callOperation how to handle more than one result.
```
I think we support two result GPRs. See extractResult etc.

kmiller68 replied:
extractResult doesn't quite work because it does a single register on 64-bit. I guess I could do:
```
setupArguments<decltype(operationEnumeratorNextUpdateIndexAndMode)>(LinkableConstant::globalObject(*this, node), baseRegs, indexGPR, modeGPR, enumeratorGPR);
appendCallSetResult(operationEnumeratorNextUpdateIndexAndMode, resultGPR1, resultGPR2);
```

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -14192,27 +14191,48 @@ IGNORE_CLANG_WARNINGS_END
             index = m_out.phi(Int32, initialIndex, incrementedIndex);
-            setJSValue(m_out.bitOr(m_out.zeroExt(index, Int64), m_out.constInt64(JSValue::DoubleEncodeOffset | static_cast<uint64_t>(JSPropertyNameEnumerator::OwnStructureMode) << 32)));
+            m_tupleValues.at(m_node->tupleOffset()) = LoweredNodeValue(index, m_highBlock);
+            m_tupleValues.at(m_node->tupleOffset() + 1) = LoweredNodeValue(m_out.constInt32(static_cast<uint32_t>(JSPropertyNameEnumerator::OwnStructureMode)), m_highBlock);
```
How about defining `setTuple(...varargs)` to do this automatically? Like,
```
setTuple(index, m_out.constInt32(static_cast<uint32_t>(JSPropertyNameEnumerator::OwnStructureMode));
```

kmiller68 replied:
Ok. Done.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -14192,27 +14191,48 @@ IGNORE_CLANG_WARNINGS_END
-
-        setInt32(m_out.castToInt32(boxedPair));
+        LValue highBits = m_out.castToInt32(m_out.lShr(boxedPair, m_out.constInt32(32)));
+        m_tupleValues.at(m_node->tupleOffset() + 1) = LoweredNodeValue(m_out.bitAnd(highBits, m_out.constInt32(JSPropertyNameEnumerator::enumerationModeMask)), m_highBlock);
```
Ditto, like,
```
setTuple(m_out.castToInt32(boxedPair), m_out.bitAnd(highBits, m_out.constInt32(JSPropertyNameEnumerator::enumerationModeMask));
```

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -14169,7 +14167,8 @@ IGNORE_CLANG_WARNINGS_END
             LValue finalIndex = m_out.phi(Int32, finalIncrementedIndex, finalPropertyIndex);
-            setJSValue(m_out.bitOr(m_out.zeroExt(finalIndex, Int64), m_out.constInt64(JSValue::NumberTag | static_cast<uint64_t>(JSPropertyNameEnumerator::IndexedMode) << 32)));
+            m_tupleValues.at(m_node->tupleOffset()) = LoweredNodeValue(finalIndex, m_highBlock);
+            m_tupleValues.at(m_node->tupleOffset() + 1) = LoweredNodeValue(m_out.constInt32(static_cast<uint32_t>(JSPropertyNameEnumerator::IndexedMode)), m_highBlock);
```
Ditto, like,
```
setTuple(finalIndex, m_out.constInt32(static_cast<uint32_t>(JSPropertyNameEnumerator::IndexedMode));
```

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/dfg/DFGVirtualRegisterAllocationPhase.cpp:
```diff
@@ -69,6 +69,13 @@ class VirtualRegisterAllocationPhase : public Phase {
                     ASSERT(!node->child1()->hasResult());
                     break;
+
+                case ExtractFromTuple:
+                    ASSERT(m_graph.m_tupleData.at(node->tupleIndex()).refCount == 1);
+                    node->setVirtualRegister(m_graph.m_tupleData.at(node->tupleIndex()).virtualRegister);
+                    ASSERT(!node->mustGenerate());
+                    continue;
```
Don't we need to free VirtualRegister when we finished using it?

kmiller68 replied:
That should happen when the ExtractFromTuple gets used. The extract is stealing the virtual register from the tuple, and we assume there's at most one extract for each tuple entry in the DFG (This isn't expected for the FTL but I'm not sure how we'd ever generate such a graph).

Constellation commented on Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
```diff
@@ -1276,7 +1299,8 @@ class PredictionPropagationPhase : public Phase {
         case EnumeratorNextUpdateIndexAndMode: {
-            setPrediction(SpecFullNumber);
+            setTuplePrediction(SpecInt32Only, 0);
+            setTuplePrediction(SpecInt32Only, 1);
```
How about defining `setTuplePrediction(SpecInt32Only, SpecInt32Only)`, taking varargs of SpeculatedType.

kmiller68 replied:
Sure.

Constellation commented on Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
```diff
@@ -1253,9 +1276,9 @@ class PredictionPropagationPhase : public Phase {
-            setPrediction(SpecInt32Only);
+        case ExtractFromTuple: {
+            // Use mergePrediction because ExtractFromTuple doesn't know if the prediction could change.
+            mergePrediction(m_tupleSpeculations[m_currentNode->tupleIndex()]);
```
How about having this prediction in TupleData? This can be useful in FixupPhase too.

kmiller68 replied:
Why would you do that here? This is the prediction on the extract node.

Constellation commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -1176,6 +1177,14 @@ class Graph final : public virtual Scannable {
     Vector<BasicBlock*, 1> m_roots;
     Vector<Edge, 16> m_varArgChildren;
 
+    struct TupleData {
+        uint16_t refCount { 0 };
+        uint16_t resultFlags { 0 };
+        VirtualRegister virtualRegister;
+    };
```
I think we can extend this further, like having m_prediction etc. Basically, this can be a data structure containing most of `DFG::Node`.

kmiller68 replied:
You don't really need the prediction here because consumers of the node will see the prediction via the ExtractFromTuple. I could see adding OpInfos here for extra metadata as needed but it seems like we should do that when we need it.

Constellation commented on Source/JavaScriptCore/dfg/DFGNode.h:
```diff
@@ -1556,6 +1556,48 @@ struct Node {
+    unsigned tupleOffset() const
+    {
+        ASSERT(isTuple());
+        return m_virtualRegister.toLocal();
```
This is a bit unfortunate, but yeah, it works since we never allocate VirtualRegister for tuple.

kmiller68 replied:
Yeah, I figured saving the OpInfos for other data seemed more useful since we know that we're not going to use the virtualRegister for anything.

Constellation commented on Source/JavaScriptCore/dfg/DFGGraph.cpp:
```diff
@@ -763,7 +763,9 @@ class RefCountCalculator {
             for (unsigned phiIndex = block->phis.size(); phiIndex--;)
                 block->phis[phiIndex]->setRefCount(0);
```
Let's extend DFGGraph.cpp to support dumping tuple results.

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16191,6 +16175,16 @@ void SpeculativeJIT::compileIdentity(Node* node)
+    ASSERT(m_graph.m_tupleData.at(node->tupleIndex()).virtualRegister == node->virtualRegister());
+    VirtualRegister virtualRegister = node->virtualRegister();
+    GenerationInfo& info = generationInfoFromVirtualRegister(virtualRegister);
+    info.stealForNode(node);
```
Maybe, we should have a bit better name for that, like, `initFromTupleElement`?
And don't we need to setup this GenerationInfo based on resultFlags of tuple element? (As the same way to FTL)
Maybe, the producer side already initialize this GenerationInfo with appropriate types, and since we are using the same VirtualRegister here, we do not need to initialize it much, but maybe we should assert here to ensure that the resulted GenerationInfo matches against the expected Tuple element's type.

kmiller68 replied:
Yeah, that's right I was relying on the fact that we are just repurposing the layout of the tuple's result. Sure, I can add those ASSERTs.

Constellation commented on Source/JavaScriptCore/b3/B3Validate.cpp:
```diff
@@ -662,6 +662,12 @@ class Validater {
                 VALIDATE(!value->kind().hasExtraBits(), ("At ", *value));
                 VALIDATE(value->numChildren() >= 1, ("At ", *value));
                 VALIDATE(value->child(0)->type() == pointerType(), ("At ", *value));
+                if (value->type().isTuple()) {
```
Let's add comment that currently we are only supporting two-element case.

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/b3/air/AirCustom.cpp:
```diff
@@ -75,23 +82,32 @@ bool CCallCustom::isValidForm(Inst& inst)
 
+    if (!inst.args[0].isSpecial())
+        return false;
+
```
Why do we need this check twice?

kmiller68 replied:
I forgot to delete it here. Will delete.

Constellation commented on Source/JavaScriptCore/b3/air/AirCustom.cpp:
```diff
@@ -75,23 +82,32 @@ bool CCallCustom::isValidForm(Inst& inst)
     // The arguments can only refer to the stack, tmps, or immediates.
-    for (Arg& arg : inst.args) {
+    for (unsigned i = inst.args.size() - 1; i > 1; --i) {
+        Arg arg = inst.args[i];
```
This does not visit `args[1]`. `i` starts with size - 1, and we run this loop if `i > 1`, this means i needs to be 2 or more. So, it misses the visit to `callee`, is there any reason?

kmiller68 replied:
Ah yeah, it should include callee. I think I confused myself with the args layout for CCallSpecial when we lower to that.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
-
-        jsValueResult(resultRegs, node);
+        useChildren(node);
+        strictInt32TupleResultWithoutUsingChildren(newIndex.gpr(), node, 0);
```
Don't we need to check `refCount` for this first tuple element?

kmiller68 replied:
We don't bother constant folding it because it's the index. If we did you'd need to check yeah but you'd trigger an assert to notice you did it wrong.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
-        JSValueRegsTemporary result(this);
+        GPRTemporary newIndex(this);
+        GPRTemporary newMode(this, Reuse, mode);
         GPRReg baseGPR = base.gpr();
```
Let's first declare all GPRs here like,
```
GPRReg newIndexGPR = newIndex.gpr();
GPRReg newModeGPR = newMode.gpr();
```
since gpr() is side-effectful operation. If we jump over that, it causes the bug. So it is always nice if we get all GPRs first.

kmiller68 replied:
I think it used to be side-effectful but `.*Temporary.gpr()` is not anymore, which IMO causes a bit of an anti pattern where the operands you declared can get spilled before they are filled. `.*Operand`s still get filled on first use so those have to be initialized early.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
-        JSValueRegsTemporary result(this);
-        GPRTemporary scratch(this, Reuse, index);
-        JSValueRegs resultRegs = result.regs();
+        GPRTemporary newIndex(this, Reuse, index);
+        GPRTemporary scratch(this);
```
Ditto for GPRs

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
+            move(TrustedImm32(static_cast<unsigned>(JSPropertyNameEnumerator::OwnStructureMode)), newMode.gpr());
+
+        useChildren(node);
+        strictInt32TupleResultWithoutUsingChildren(newIndex.gpr(), node, 0);
```
Ditto.

kmiller68 replied:
Ditto, not constant folded.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
+#if USE(JSVALUE64)
+    GPRTemporary newMode(this, Reuse, mode);
+#endif
     JSValueRegs baseRegs = base.regs();
```
Ditto. Let's get all GPRs here first.

kmiller68 replied:
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -14482,118 +14482,86 @@ void SpeculativeJIT::compileEnumeratorNextUpdateIndexAndMode(Node* node)
 
-    strictInt32Result(resultGPR, node);
+    useChildren(node);
+    strictInt32TupleResultWithoutUsingChildren(indexResult.gpr(), node, 0);
+    strictInt32TupleResultWithoutUsingChildren(modeResult.gpr(), node, 1);
```
Don't we need to check refCount?

kmiller68 replied:
I think I'll just change it so we check in `strictInt32TupleResultWithoutUsingChildren`.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16191,6 +16159,42 @@ void SpeculativeJIT::compileIdentity(Node* node)
+    ASSERT(m_graph.m_tupleData.at(node->tupleIndex()).virtualRegister == node->virtualRegister());
+    VirtualRegister virtualRegister = node->virtualRegister();
+    GenerationInfo& info = generationInfoFromVirtualRegister(virtualRegister);
+    if (ASSERT_ENABLED) {
```
IIRC, this causes some build failures. Let's just use `#if ASSERT_ENABLED`

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/b3/air/AirCustom.cpp:
```diff
@@ -65,8 +66,14 @@ bool CCallCustom::isValidForm(Inst& inst)
+
+    size_t resultCount = cCallResultCount(code, value);
+    size_t expectedArgCount = resultCount + 1; // first Arg is always CCallSpecial.
     for (Value* child : value->children()) {
```
OK, it is including callee itself.

Constellation commented on Source/JavaScriptCore/b3/air/AirCustom.h:
```diff
@@ -134,19 +134,23 @@ struct CCallCustom : public CommonCustomBase<CCallCustom> {
 
-        for (size_t n = cCallResultCount(value); n; --n) {
+        for (size_t n = cCallResultCount(code, value); n; --n) {
+            Type type = value->type().isTuple() ? code.proc().typeAtOffset(value->type(), n - 1) : value->type();
```
`n` is decremented. But `next`'s `index` is incremented. Is this correct?

kmiller68 replied:
I guess it doesn't matter right now because if there's more than one result you get the same registers anyway. In the future it could matter though so I'll fix.

Constellation reviewed (approved):
r=me

kmiller68 committed:
DFG should support tuples
https://bugs.webkit.org/show_bug.cgi?id=253413

Reviewed by Yusuke Suzuki.

This change adds support for tuples in the DFG. It works similarly to how tuples
work in B3 where there is a side buffer which holds most of the metadata for a
tuple. In the DFG there are three pieces of information held here:
    1) The reference count
    2) The result flags
    3) The virtual register

1) tells us if the ExtractFromTuple Node for a given tuple result still exists
(e.g. it could have been constant folded or dead code eliminated). From that we
will decide to fill (3) the virtual register when allocating registers. If we
didn't have the reference count then we would have no way to know that the
virtual registers isn't going to be consumed thus we will leak it. Lastly we
have (2) the result flags for the node. This tells the ExtractFromTuple and
theoretically (as right now the only ExtractFromTuple users produce Int32s) any
 consumers of the ExtractFromTuple what value format the tuple's result will
 be in. For the DFG, we don't need this because we know there's at most one
 ExtractFromTuple for any given tuple index since we don't duplicate code.

When dumping the DFG graph ExtractFromTuple follows the same pattern as B3
and uses `<<X` to denote the offset we are extracting from. So it will look
something like:

  6  1  0:  D@101:< 1:->    EnumeratorNextUpdateIndexAndMode(Check:Untyped:D@97, Check:Untyped:D@98, Check:Untyped:D@100, Check:Untyped:D@99, VarArgs, SelectUsingPredictions+NonArray+InBounds+AsIs+Read, enumeratorModes = 4, R:World, W:Heap, Exits, ClobbersExit, bc#115, ExitValid)
  7  1  0:  D@102:< 1:->    ExtractFromTuple(Check:Untyped:D@101, Int32|UseAsOther, <<0, bc#115, ExitInvalid)
  8  1  0:  D@103:<!0:->    MovHint(Check:Untyped:D@102, MustGen, loc10, W:SideState, ClobbersExit, bc#115, ExitInvalid)
  9  1  0:  D@104:< 1:->    ExtractFromTuple(Check:Untyped:D@101, Int32|UseAsOther, <<1, bc#115, ExitInvalid)

This patch also adds support for calling operations in both the FTL/B3 via
CCall. CCall can take exactly the tuple of `{ pointerType(), pointerType() }`,
which, for every calling conevention we support, should be returned in both
the return value registers. As the only way to look into a tuple is via the
B3 prodecure, the first Air::Arg of any CCall/ColdCCall Inst is now the
CCallSpecial for the compiling Air::Code. This gives us access to Air::Code
inside CCallCustom::forEachArg and isValidForm.

* JSTests/stress/for-in-redefine-enumerable.js:
(shouldBe):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Type.h:
(JSC::B3::pointerType):
(JSC::B3::registerType):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:
(JSC::B3::Air::cCallResultCount):
(JSC::B3::Air::cCallArgumentRegisterWidth):
(JSC::B3::Air::cCallResult):
* Source/JavaScriptCore/b3/air/AirCCallingConvention.h:
* Source/JavaScriptCore/b3/air/AirCustom.cpp:
(JSC::B3::Air::CCallCustom::isValidForm):
* Source/JavaScriptCore/b3/air/AirCustom.h:
(JSC::B3::Air::CCallCustom::forEachArg):
* Source/JavaScriptCore/b3/air/AirLowerAfterRegAlloc.cpp:
(JSC::B3::Air::lowerAfterRegAlloc):
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_3.cpp:
(addCallTests):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(JSC_DEFINE_JIT_OPERATION):
(testCallPairResult):
(testCallPairResultRare):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreter.h:
(JSC::DFG::AbstractInterpreter::setTupleConstant):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGAtTailAbstractState.cpp:
(JSC::DFG::AtTailAbstractState::AtTailAbstractState):
* Source/JavaScriptCore/dfg/DFGAtTailAbstractState.h:
(JSC::DFG::AtTailAbstractState::forTupleNode):
(JSC::DFG::AtTailAbstractState::clearForTupleNode):
(JSC::DFG::AtTailAbstractState::setForTupleNode):
(JSC::DFG::AtTailAbstractState::setTypeForTupleNode):
(JSC::DFG::AtTailAbstractState::setNonCellTypeForTupleNode):
(JSC::DFG::AtTailAbstractState::makeBytecodeTopForTupleNode):
(JSC::DFG::AtTailAbstractState::makeHeapTopForTupleNode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::addToGraph):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGenerationInfo.h:
(JSC::DFG::GenerationInfo::initFromTupleResult):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::dump):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.cpp:
(JSC::DFG::InPlaceAbstractState::InPlaceAbstractState):
* Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.h:
(JSC::DFG::InPlaceAbstractState::forTupleNode):
(JSC::DFG::InPlaceAbstractState::clearForTupleNode):
(JSC::DFG::InPlaceAbstractState::setForTupleNode):
(JSC::DFG::InPlaceAbstractState::setTypeForTupleNode):
(JSC::DFG::InPlaceAbstractState::setNonCellTypeForTupleNode):
(JSC::DFG::InPlaceAbstractState::makeBytecodeTopForTupleNode):
(JSC::DFG::InPlaceAbstractState::makeHeapTopForTupleNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::isTuple const):
(JSC::DFG::Node::setTupleOffset):
(JSC::DFG::Node::tupleOffset const):
(JSC::DFG::Node::hasExtractOffset const):
(JSC::DFG::Node::extractOffset const):
(JSC::DFG::Node::tupleIndex const):
(JSC::DFG::Node::tupleSize const):
(JSC::DFG::Node::hasVirtualRegister):
(JSC::DFG::Node::virtualRegister):
(JSC::DFG::Node::setVirtualRegister):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
(JSC::DFG::makeUGPRPair):
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGScoreBoard.h:
(JSC::DFG::ScoreBoard::use):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::strictInt32TupleResultWithoutUsingChildren):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGValidate.cpp:
* Source/JavaScriptCore/dfg/DFGVirtualRegisterAllocationPhase.cpp:
(JSC::DFG::VirtualRegisterAllocationPhase::run):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::LowerDFGToB3):
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::extract):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::AirIRGeneratorBase::emitCCall):

Canonical link: https://commits.webkit.org/262068@main
